Added internal solution to cf 1051F problem (#6098)
cpinitiative/usaco-guide: solutions/gold/cf-1051F.mdx

---
id: cf-1051F
source: CF
title: Euler Tour Technique
author: Raagav Bala
---

[Official Analysis (C++)](https://codeforces.com/blog/entry/61969)

## Explanation

> First, notice that $m - n \le 20$ suggests that the graph is almost a tree, except a few extra edges in addition to the tree.

Reviewer: immediately after you talk about tree stuff so I think you might need to segue more, e.g.

> First, since the graph is connected and the number of edges $m$ is bounded by $n + 20$, we can consider it as a spanning tree with a few extra edges added. Since the number of extra edges is fairly small, we can first process the main tree, and determine how to handle the extra edges later.

> We can take advantage of this by doing a DFS from the root, to find each node's depth and its distance from the root, and so that we can create the binary lifting table for LCA. Because every edge we take in the DFS is part of the tree, we can remove it from the set of edges not yet visited in the DFS, where we denote as `unused`. After the DFS, the edges in `unused` are not part of the tree, and there are only at most $m - n + 1 \le 21$ of them.

Reviewer 2: ```suggestion
We can take advantage of this by running a DFS from the root to find each node's depth and distance from the root, and to build a binary lifting table for LCA. Because every edge we consider in the DFS is part of the tree, we can remove it from the set of unvisited edges, which we denote as `unused`. After the DFS, the `unused` edges aren't part of the tree anymore, and so there are only at most $m - n + 1 \le 21$ of them.
```

> So, the shortest distance between two nodes $u$ and $v$ using only the nodes in the tree can be calculated using LCA:

Reviewer 2: ```suggestion
Therefore, the shortest distance between two nodes $u$ and $v$ using only the tree's nodes can be calculated using LCA:
```
